### PR TITLE
enable mTLS from proxyless gRPC clients

### PR DESCRIPTION
**Pull request by stevenctl**
* Set the experimental flag
* Build proper transport credential
* Cleanup CDS
* Cleanup bootstrap:
  * correct plugin name
  * properly marshal durationpb  

---

Tested with:

```
apiVersion: security.istio.io/v1beta1
kind: PeerAuthentication
metadata:
  name: "default"
spec:
  mtls:
    mode: STRICT
```

Before, we'd not even attempt TLS and the request would fail. This change gets me a successful request to a workload running envoy, with TLS. 

Screen Shot 2021-07-16 at 12 18 46 PM

---

Planning to add proxyless to the security reachabilty suite in a follow up

### CODE REVIEW

**Comment by howardjohn** [comment]
How does auto mtls work to grpc? I assume we cannot tell if it will accept TLS or not since *grpc ignores the TLS configuration sent over XDS*

File: pkg/istio-agent/grpcxds/grpc_bootstrap.go [hunk 1]
```diff
@@ -51,17 +52,17 @@ type XdsServer struct {
...
 
 type FileWatcherCertProviderConfig struct {
-	CertificateFile   string               `json:"certificate_file,omitempty"`
-	PrivateKeyFile    string               `json:"private_key_file,omitempty"`
-	CACertificateFile string               `json:"ca_certificate_file,omitempty"`
-	RefreshDuration   *durationpb.Duration `json:"refresh_interval,omitempty"`
+	CertificateFile   string          `json:"certificate_file,omitempty"`
+	PrivateKeyFile    string          `json:"private_key_file,omitempty"`
+	CACertificateFile string          `json:"ca_certificate_file,omitempty"`
+	RefreshDuration   json.RawMessage `json:"refresh_interval,omitempty"`
```
Comment (howardjohn) [inline pkg/istio-agent/grpcxds/grpc_bootstrap.go:65]: does just time.Duration work?

**Review by howardjohn — APPROVED** [review]

File: pkg/istio-agent/grpcxds/grpc_bootstrap.go (hunk 1 above)
Comment (stevenctl) [reply pkg/istio-agent/grpcxds/grpc_bootstrap.go:65]: Possibly? I rewrote this to match their unmarshaling

Comment (howardjohn) [reply pkg/istio-agent/grpcxds/grpc_bootstrap.go:65]: I think it would but matching there's is better

**Comment by stevenctl** [comment]
> grpc ignores the TLS configuration sent over XDS

Do you mean for gRPC servers? That has to be dealt with separately, and it also requires application code changes to use https://pkg.go.dev/google.golang.org/grpc/xds#NewGRPCServer.

Clients won't ignore it if the flag is set and right client code is used, and without the peerAuthentication this change causes failures, not sure why exactly; I figured even without STRICT the server-side envoy would still perform TLS. 

```
WARNING: 2021/07/17 17:01:41 [core] grpc: addrConn.createTransport failed to connect to {10.120.1.14:17070 a.default.svc.cluster.local:7070 0xc00496a550 0 <nil>}. Err: connection error: desc = "transport: authentication handshake failed: read tcp 10.120.1.17:48262->10.120.1.14:17070: read: connection reset by peer". Reconnecting...
```

If you mean how do we know if the client will actually use the TLS configuration we send it... that I'm not sure. If its client-side, things will probably just break for the most part. For servers ignoring TLS config we should definitely find a way to raise a warning. 

I'll update this [doc](https://docs.google.com/document/d/1UG-s2mG-Sjt9_OzTLFPc-WJiZ09ins48kS5lnMD2V6o/edit?usp=sharing) as I figure more things out.

**Comment by howardjohn** [comment]
> Do you mean for gRPC servers? 

What I mean is in k8s, we know if the server will accept mTLS traffic by logic which is basically `hasSidecar && !PeerAuth=DISABLE`. This works in almost all cases (one case it does not is multiprimary - we don't know which PeerAuth is used).

For grpc its a bit different. First, we do not have an equivalent `hasSidecar` check we can do. You could kind of do one from injection but not really a good idea. Even if they are using XDS, they may still not have the tls config set in their application, so the presences of PeerAuth is not sufficient as they may just ignore it.

So I am not sure we have a good way to implement this core functionality

**Comment by stevenctl** [comment]
By defaulting everything to plaintext, and only claiming support for STRICT mode, if either side is miconfigured requests will break rather than a user expecting mTLS but actually doing plaintext... ? Agree we don't have a good way to implement this.

**Pull request merged**